api/src/test/java/org/openmrs/api/VisitServiceTest.java

```diff
@@ -966,11 +978,222 @@ public void saveVisit_shouldNotPersistNewEncounter() {
 	public void getAllVisitTypes_shouldGetAllVisitTypesBasedOnIncludeRetiredFlag() {
 		VisitService visitService = Context.getVisitService();
 		List<VisitType> visitTypes = visitService.getAllVisitTypes(true);
-		assertEquals(3, visitTypes.size(), "get all visit types including retired");
+		assertEquals(4, visitTypes.size(), "get all visit types including retired");
 		visitTypes = visitService.getAllVisitTypes(false);
-		assertEquals(2, visitTypes.size(), "get all visit types excluding retired");
+		assertEquals(3, visitTypes.size(), "get all visit types excluding retired");
 	}
-
+	
+	/**
+	 * @see org.org.openmrs.api.VisitService#ensureVisit(Patient, Date, Location)
+	 */
+	@Test
+	public void ensureVisit_shouldCreatesNewVisitIfVisitTimeIsEitherLessThanVisitStartTimeOrGreaterThanVisitEndTime() {
+		Location visitLocation = new Location(1901);
+		visitLocation.addTag(supportsVisits);
+		Patient patient = new Patient(6);
+		assertFalse(visitService.getVisitsByPatient(patient).isEmpty(), "patients has no visits");
+		//ended ten days ago
+		Date now = new Date();
+		Date startTime = DateUtils.addDays(now, -10);
+		
+		Visit newVisit = Context.getVisitService().getVisit(4);
+		newVisit.setLocation(visitLocation);
+		newVisit.setStartDatetime(DateUtils.addDays(now, -4));
+		newVisit.setStopDatetime(DateUtils.addDays(now, -2));
+		newVisit.setPatient(patient);
+		
+		Visit oldVisit = Context.getVisitService().getVisit(5);
+		oldVisit.setLocation(visitLocation);
+		oldVisit.setStartDatetime(DateUtils.addDays(now, -6));
+		oldVisit.setStopDatetime(DateUtils.addDays(now, -1));
+		oldVisit.setPatient(patient);
+		
+		int originalVisits = Context.getVisitService().getAllVisits().size();
+		
+		final Visit returnedVisit = visitService.ensureVisit(patient, startTime, visitLocation);
+		assertNotNull(returnedVisit);
+		assertNotSame(newVisit, returnedVisit);
+		assertThat(Context.getVisitService().getAllVisits().size(), is(originalVisits + 1));
+	}
+	
+	/**
+	 * @see org.org.openmrs.api.VisitService#isSuitableVisit(Visit, Location, Date)
+	 */
+	@Test
+	public void isSuitableVisit_shouldReturnFalseIfWhenIsEitherLessThanVisitStartTimeOrGreaterThanVisitEndTime() {
```

Comment:
Whatdo you mean by `shouldReturnFalseIfWhenIsEitherLessThanVisitStartTimeOrGreaterThanVisitEndTime` ?

Reply:
@tendomart what i mean is when should be in the range `visitStartTime` < `when` < `visitEndTime` and return false otherwise.  How best can we represent this

Reply:
Okk perfect